cmd/docker-proxy/tcp_proxy.go

```diff
@@ -35,52 +35,44 @@ func NewTCPProxy(frontendAddr, backendAddr *net.TCPAddr) (*TCPProxy, error) {
 	}, nil
 }
 
-func (proxy *TCPProxy) clientLoop(client *net.TCPConn, quit chan bool) {
-	backend, err := net.DialTCP("tcp", nil, proxy.backendAddr)
+func (proxy *TCPProxy) clientLoop(client *net.TCPConn) {
+	conn, err := net.DialTimeout("tcp", proxy.backendAddr.String(), 5*time.Second)
 	if err != nil {
-		log.Printf("Can't forward traffic to backend tcp/%v: %s\n", proxy.backendAddr, err)
 		client.Close()
+		log.Printf("Can't forward traffic to backend tcp/%v: %s\n", proxy.backendAddr, err)
 		return
 	}
+	backend := conn.(*net.TCPConn)
+
+	done := make(chan struct{}, 2)
 
-	var wg sync.WaitGroup
 	broker := func(to, from *net.TCPConn) {
 		io.Copy(to, from)
-		from.CloseRead()
-		to.CloseWrite()
-		wg.Done()
+		done <- struct{}{}
 	}
 
-	wg.Add(2)
 	go broker(client, backend)
 	go broker(backend, client)
 
-	finish := make(chan struct{})
-	go func() {
-		wg.Wait()
-		close(finish)
-	}()
+	<-done
 
-	select {
-	case <-quit:
-	case <-finish:
-	}
-	client.Close()
+	// reduce close_wait states
+	backend.SetLinger(0)
+	backend.SetDeadline(time.Now().Add(time.Millisecond))
```

Comment:
This is not safe to do in a general-purpose TCP proxy. https://vincent.bernat.ch/en/blog/2014-tcp-time-wait-state-linux#socket-lingering